pkg/runtime/dispatcher.go

```diff
@@ -0,0 +1,443 @@
+	// ready to use handler
+	handler adapter.Handler
+	// configuration for the handler
+	handlerConfig *cpb.Handler
+	// instanceConfigs to dispatch to the handler.
+	// instanceConfigs must belong to the same template.
+	instanceConfig []*cpb.Instance
+}
+
+// Resolver represents the current snapshot of the configuration database
+// and associated, initialized handlers.
+type Resolver interface {
+	// Resolve resolves configuration to a list of actions.
+	Resolve(bag attribute.Bag, variety adptTmpl.TemplateVariety) ([]*Action, error)
+}
+
+// Newdispatcher creates a new dispatcher.
+func Newdispatcher(mapper expr.Evaluator, templateRepo template.Repository,
+	rt Resolver, gp *pool.GoroutinePool) Dispatcher {
+	m := &dispatcher{
+		mapper:       mapper,
+		templateRepo: templateRepo,
+		gp:           gp,
+	}
+	m.SetResolver(rt)
+	return m
+}
+
+// dispatcher is responsible for dispatching incoming API calls
+// to the configured adapters. It implements the Dispatcher interface.
+type dispatcher struct {
+	// mapper is the selector and expression evaluator.
+	// It is not directly used by dispatcher.
+	mapper expr.Evaluator
+
+	// Repository of templates that are compiled in this Mixer.
+	// It provides a way to perform template operations.
+	templateRepo template.Repository
+
+	// <resolver>
+	resolver atomic.Value
+
+	// gp is used to dispatch multiple adapters concurrently.
+	gp *pool.GoroutinePool
+}
+
+// SetState installs a new runtime state.
+func (m *dispatcher) SetResolver(rt Resolver) {
+	m.resolver.Store(rt)
+}
+
+// State gets the current runtime state.
+func (m *dispatcher) Resolver() Resolver {
+	return m.resolver.Load().(Resolver)
+}
+
+// Dispatcher dispatches incoming API calls to configured adapters.
+type Dispatcher interface {
+
+	// Preprocess dispatches to the set of aspects that will run before any
+	// other aspects in Mixer (aka: the Check, Report, Quota aspects).
+	Preprocess(ctx context.Context, requestBag attribute.Bag, responseBag *attribute.MutableBag) error
+
+	// Check dispatches to the set of aspects associated with the Check API method
+	Check(ctx context.Context, requestBag attribute.Bag) (*adapter.CheckResult, error)
+
+	// Report dispatches to the set of aspects associated with the Report API method
+	Report(ctx context.Context, requestBag attribute.Bag) error
+
+	// Quota dispatches to the set of aspects associated with the Quota API method
+	Quota(ctx context.Context, requestBag attribute.Bag,
+		qma *aspect.QuotaMethodArgs) (*adapter.QuotaResult2, error)
+}
+
+// dispatchInfo contains details needed to make a function call
+type dispatchInfo struct {
+	// generated code - processor.ProcessXXX ()
+	processor *template.Info
+	// Name of the handler being called. Informational.
+	handlerName string
+	// Name of adapter that created the handler. Informational.
+	adapterName string
+	// handler to call.
+	handler adapter.Handler
+	// instance configuration to call the handler with.
+	instances []*cpb.Instance
```

Comment:
Similar field is called instanceConfig in the Action struct, t'would be good to be consistent.